File: waste_oil_backend/apps/accounts/permissions.py
from rest_framework.permissions import BasePermission

from .models import CustomUser


def _role(request):
    user = getattr(request, "user", None)
    if not user or not user.is_authenticated:
        return None
    return getattr(user, "role", None)


class IsStoreman(BasePermission):
    def has_permission(self, request, view):
        return _role(request) == CustomUser.Role.STOREMAN


class IsTreatment(BasePermission):
    def has_permission(self, request, view):
        return _role(request) == CustomUser.Role.TREATMENT


class IsAdminDept(BasePermission):
    def has_permission(self, request, view):
        return _role(request) == CustomUser.Role.ADMIN


class IsManager(BasePermission):
    def has_permission(self, request, view):
        return _role(request) == CustomUser.Role.MANAGER


class IsGM(BasePermission):
    def has_permission(self, request, view):
        return _role(request) == CustomUser.Role.GM


class IsStoremanGmOrSuperadmin(BasePermission):
    """Storeman, GM, or Superadmin — manage vendor master data."""

    def has_permission(self, request, view):
        return _role(request) in (
            CustomUser.Role.STOREMAN,
            CustomUser.Role.GM,
            CustomUser.Role.SUPERADMIN,
        )


class IsGMOrSuperadmin(BasePermission):
    """General Manager or Superadmin — employee & department management."""

    def has_permission(self, request, view):
        return _role(request) in (
            CustomUser.Role.GM,
            CustomUser.Role.SUPERADMIN,
        )


class IsManagerOrAbove(BasePermission):
    def has_permission(self, request, view):
        return _role(request) in (
            CustomUser.Role.MANAGER,
            CustomUser.Role.GM,
            CustomUser.Role.SUPERADMIN,
        )


class IsCurrentHolder(BasePermission):
    message = "Only the current holder can perform this action."

    def has_object_permission(self, request, view, obj):
        user = getattr(request, "user", None)
        if not user or not user.is_authenticated:
            return False
        holder_id = getattr(obj, "current_holder_id", None)
        return holder_id is not None and holder_id == user.id

File: waste_oil_desktop-TAURI/src/utils/permissions.js
export const ROLE_STAGE = {
  storeman: 1,
  treatment: 2,
  admin: 3,
  manager: 4,
  gm: 5,
  superadmin: null,
};

export function stageForRole(role) {
  const v = ROLE_STAGE[role];
  return v === undefined ? null : v;
}

export function canForward(role, currentStage) {
  const rs = stageForRole(role);
  return rs != null && rs === currentStage;
}

export function canReturn(role, currentStage) {
  return canForward(role, currentStage) && currentStage > 1;
}

export function canEdit(role, currentStage) {
  return canForward(role, currentStage);
}

export function isCurrentHolder(record, user) {
  if (!record || !user) return false;
  const hid = record.current_holder_id;
  if (hid == null) return false;
  return String(hid) === String(user.id);
}

export function canActForward(record, user) {
  if (!record || record.is_locked) return false;
  return (
    canForward(user?.role, record.current_stage) &&
    isCurrentHolder(record, user)
  );
}

export function canActReturn(record, user) {
  if (!record || record.is_locked) return false;
  return (
    canReturn(user?.role, record.current_stage) &&
    isCurrentHolder(record, user)
  );
}

export function canActEdit(record, user) {
  if (!record || record.is_locked) return false;
  return (
    canEdit(user?.role, record.current_stage) &&
    isCurrentHolder(record, user)
  );
}
